### PR TITLE
Add aws support in gitops v2

### PR DESCRIPTION
**Pull request by NissesSenap**
* The default new only uses Azure, no need for current users to change anything
* Solves #75

### CODE REVIEW

**Comment by NissesSenap** [comment]
If you think the yaml looks okay I will update the docs as well.

**Comment by simongottschlag** [comment]
I would prefer it if we could keep as much as possible DRY - but only if it's not too much work.

I would extract the `docker login` part from Azure and AWS to separate steps and load them based on what provider is being used. The rest could most likely be run with the usual `bash` task.

Example:

`main.yaml`

```YAML
trigger:
- main

pool:
  vmImage: ubuntu-latest

parameters:
- name: cloudProvider
  displayName: "Name of the cloud provider"
  type: string
  default: "azure"

steps:
- ${{ if eq(parameters.cloudProvider, 'azure') }}:
  - template: azure.yaml
    parameters:
      serviceConnection: "azure"
- ${{ if eq(parameters.cloudProvider, 'aws') }}:
  - template: aws.yaml
    parameters:
      serviceConnection: "aws"
      regionName: "eu-north-1"
```

`azure.yaml`:

```YAML
parameters:
- name: serviceConnection
  displayName: "Name of the service connection"
  type: string

steps:
- task: AzureCLI@2
  inputs:
    scriptType: bash
    scriptLocation: inlineScript
    azureSubscription: ${{ parameters.serviceConnection }}
    inlineScript: |
      echo "azure"
```

`aws.yaml`:

```YAML
parameters:
- name: serviceConnection
  displayName: "Name of the service connection"
  type: string
- name: regionName
  displayName: "Name of the AWS region"
  type: string

steps:
- task: AWSShellScript@1
  inputs:
    awsCredentials: ${{ parameters.serviceConnection }}
    regionName: ${{ parameters.regionName }}
    scriptType: 'inline'
    inlineScript: |
      echo "aws"
```

**Comment by NissesSenap** [comment]
As I wrote in the issue i realized that the problem with the serviceConnection variable was that I used the same for Azure & AWS and thanks to that the pipeline verification said no in the Azure CLI stage when working with a AWS service connection.

So I was able to wrap everything in to the push file again just as you said just run the login commands separately and store the tokens in variables and then push the images in the same task.

There is some minor changes in the main.yaml but I don't think that should create any issues for Azure users.
I will also write some docs around this in a new commit. When we think everything looks okay we squash it.

File: gitops-v2/new/push.yaml [hunk 1]
```diff
@@ -71,38 +82,94 @@ jobs:
...
+          # Run AWS
+          if [[ ${{ parameters.awsRegion }} != "" ]]
+          then
+            docker login -u AWS -p $TOKEN $SERVER
+            docker tag $(imageId) ${SERVER}:${SHA_SHORT}
+            docker push ${SERVER}:${SHA_SHORT}
+          fi
+
+          # Run Azure
+          if [[ $IMAGE_PATH_PREFIX != "" ]]
```
Comment (NissesSenap) [inline gitops-v2/new/push.yaml:159]: The reason why I use IMAGE_PATH_PREFIX is that I got issues with the templating things from main.yaml so I just took a variable that would always be a simple string.

Same thing with awsRegion.

**Comment by NissesSenap** [comment]
Updated after feedback from @simongottschlag, of course the tasks is run on the same server so I don't have to move around the docker tokens but instead just login.

File: gitops-v2/new/push.yaml [hunk 2]
```diff
@@ -71,38 +82,85 @@ jobs:
...
           scriptLocation: inlineScript
           scriptType: bash
           inlineScript: |
-            set -ex
+            set -e
 
-            WORKSPACE=$(Pipeline.Workspace)
             SERVICE_NAME=$(resources.triggeringAlias)
             IMAGE_ID=$(imageId)
             IMAGE_PATH_PREFIX=${{ parameters.imagePathPrefix }}
```
Comment (simongottschlag) [inline gitops-v2/new/push.yaml:87]: Are these still necessary here?

File: gitops-v2/new/push.yaml [hunk 3]
```diff
@@ -71,38 +82,85 @@ jobs:
...
-            EXTRA_TAG=${{ parameters.extraTag }}
-
-            ARTIFACT_PATH="${WORKSPACE}/${SERVICE_NAME}/${ARTIFACT_FILE_NAME}"
-            if [[ ! -f "${ARTIFACT_PATH}" ]]
-            then
-                echo "Missing artifact file ($ARTIFACT_PATH) from CI pipeline, did you trigger this pipeline manually?"
-                exit 1
-            fi
+
             IMAGE_ID_FULL="${ACR_NAME}.azurecr.io/${IMAGE_PATH_PREFIX}/${IMAGE_ID}"
```
Comment (simongottschlag) [inline gitops-v2/new/push.yaml:97]: Is this one needed here?

File: gitops-v2/new/push.yaml [hunk 4]
```diff
@@ -71,38 +82,85 @@ jobs:
...
+        condition: ne('${{ parameters.awsRegion }}', '')
+
+      - bash: |
+          set -e
+          WORKSPACE=$(Pipeline.Workspace)
+          EXTRA_TAG=${{ parameters.extraTag }}
+          SHA_SHORT=$(shaShort)
+          SERVER=$(server)
+          ACR_NAME=${{ parameters.acrName }}
+          IMAGE_PATH_PREFIX=${{ parameters.imagePathPrefix }}
```
Comment (simongottschlag) [inline gitops-v2/new/push.yaml:131]: While we are at it, I think we should move these to the task ENV instead. No need to do it inside of the task from what I see here.

File: gitops-v2/new/push.yaml [hunk 5]
```diff
@@ -71,38 +82,85 @@ jobs:
...
+          IMAGE_ID_FULL="${ACR_NAME}.azurecr.io/${IMAGE_PATH_PREFIX}/${IMAGE_ID}"
```
Comment (simongottschlag) [inline gitops-v2/new/push.yaml:132]: If we export this one from both Azure and AWS, then maybe we wouldn't need to have an if statement in bash for the providers?

File: gitops-v2/new/push.yaml [hunk 6]
```diff
@@ -71,38 +82,85 @@ jobs:
...
-              docker push ${SERVICE_NAME}:${EXTRA_TAG}
-            fi
+          fi
+
+          if [[ "${EXTRA_TAG}" != "" ]]
+          then
+            docker tag ${SERVICE_NAME}:${EXTRA_TAG}
+            docker push ${SERVICE_NAME}:${EXTRA_TAG}
+          fi
+
```
Comment (simongottschlag) [inline gitops-v2/new/push.yaml:162]: I actually don't see how this one works. No registry to push to.. ?

**Review by simongottschlag — CHANGES_REQUESTED** [review]
I think the task would be easier to understand with a few changes:
- Export the remote tag from AWS/Azure and use it in the push instead of an if statement
- Where we don't do anything in bash with variables, use env
- Remove unused variables

File: gitops-v2/new/push.yaml (hunk 6 above)
Comment (NissesSenap) [reply gitops-v2/new/push.yaml]: Me nether to be honest, I guess it works if you have a repo that is public for everyone to write too :)
I just stayed consistent with the current solution. I will remove them both

File: gitops-v2/new/push.yaml (hunk 2 above)
Comment (NissesSenap) [reply gitops-v2/new/push.yaml]: Nope i will remove.

**Review by simongottschlag — APPROVED** [review]

File: gitops-v2/new/push.yaml (hunk 5 above)
Comment (NissesSenap) [reply gitops-v2/new/push.yaml]: Fixed, maybe not the most beautiful solution but it works at least :)

**Pull request merged**